rewriter/ChainedSig.cc

```diff
@@ -0,0 +1,178 @@
+    // `sig` -> error no block
+    // `sig.final {}` -> no error. Need to prevent adding the error when we receive `sig` by itself
+    ast::Send *previousSigSend;
+
+    ast::ExpressionPtr preTransformSend(core::MutableContext ctx, ast::ExpressionPtr tree) {
+        auto *send = ast::cast_tree<ast::Send>(tree);
+        ast::Send *sigSend = ast::cast_tree<ast::Send>(send->recv);
+
+        // Make sure the first receiver is sig. E.g.: `sig.abstract {}`, `sig.override.final {}`
+        if (send->fun != core::Names::sig() && (sigSend == nullptr || !firstReceiverIsSig(sigSend))) {
+            return tree;
+        }
+
+        // Return early if we identify things like `params` being invoked on `sig`
+        if (invalidChainedStatement(ctx, send)) {
+            return tree;
+        }
+
+        if (send->fun == core::Names::sig() && !send->hasBlock() && send != previousSigSend) {
+            if (auto e = ctx.beginError(send->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Signature declarations expect a block");
+                e.addErrorNote("Complete the signature by adding a block declaration: sig `{}`", "{ ... }");
+            }
+
+            return tree;
+        }
+
+        // Return early unless it's one of the sends we are interested in
+        if (send->fun != core::Names::abstract() && send->fun != core::Names::final_() &&
+            send->fun != core::Names::override_() && send->fun != core::Names::overridable()) {
+            return tree;
+        }
+
+        // This is still an invocation where `sig` is the first receiver, but there is still no block. We need to pop
+        // the last entry for incomplete sigs and add the current send to account for multiple methods chained on sig
+        //
+        // E.g.: `sig.override.final {}`
+        //
+        // If we never find a send where the first receiver is a `sig` and the last invocation has the declaration
+        // block, then we add the error later on
+        if (!send->hasBlock()) {
+            if (auto e = ctx.beginError(send->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Signature declarations expect a block");
+                e.addErrorNote("Complete the signature by adding a block declaration: sig `{}`", "{ ... }");
+            }
+
+            return tree;
+        }
+
+        // For all other cases, we have to re-write the block
+        // E.g.: `sig.abstract { void }` -> `sig { abstract.void }`
+        auto *block = send->block();
+        auto blockBody = ast::cast_tree<ast::Send>(block->body);
+
+        // If the blockBody is not a send, then we have a sequence of instructions inside the signature block
+        if (blockBody == nullptr) {
+            if (auto e = ctx.beginError(block->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Malformed signature: cannot have multiple instructions inside a signature block");
```

Comment:
I think we should use "expressions" rather "instructions" in the comment here, since that corresponds to the terminology of the AST more closely.  ("instructions" are a CFG thing in Sorbet.)

I don't know what the best terminology to use in the actual error message is; I think "statements" is the best word to use.  (I think I could be convinced that "expressions" is the right thing...maybe.)